File: master_wallet/extensions/firefox_extension/services/TaxAnalyticsService.js
// TigerWallet MasterWallet - Tax Analytics Service (Chrome Extension)
// Tax reporting and analytics
// Production-ready

class MasterTaxAnalyticsService {
  constructor(masterWalletId) {
    this.masterWalletId = masterWalletId;
    this.transactions = [];
    this.reports = [];
    this.config = {
      method: 'FIFO',
      jurisdiction: 'US',
      shortTermRate: 0.37,
      longTermRate: 0.20,
      incomeTaxRate: 0.22,
    };
    this.isInitialized = false;
  }

  async initialize() {
    if (this.isInitialized) return true;
    
    try {
      // Load transactions
      await this.loadTransactions();
      
      // Load config
      await this.loadConfig();
      
      this.isInitialized = true;
      return true;
    } catch (error) {
      console.error('TaxAnalyticsService initialization failed:', error);
      return false;
    }
  }

  async loadTransactions() {
    const result = await chrome.storage.local.get('taxTransactions');
    if (result.taxTransactions) {
      this.transactions = result.taxTransactions;
    }
  }

  async loadConfig() {
    const result = await chrome.storage.local.get('taxConfig');
    if (result.taxConfig) {
      this.config = result.taxConfig;
    }
  }

  async saveTransactions() {
    await chrome.storage.local.set({
      taxTransactions: this.transactions,
    });
  }

  async saveConfig() {
    await chrome.storage.local.set({
      taxConfig: this.config,
    });
  }

  // Add transaction
  async addTransaction(tx) {
    this.transactions.push({
      ...tx,
      addedAt: Date.now(),
    });
    
    await this.saveTransactions();
    return true;
  }

  // Add multiple transactions
  async addTransactions(txs) {
    for (const tx of txs) {
      this.transactions.push({
        ...tx,
        addedAt: Date.now(),
      });
    }
    
    await this.saveTransactions();
    return true;
  }

  // Get transactions
  async getTransactions(walletAddress, filters = {}) {
    let result = this.transactions.filter(t => t.walletAddress === walletAddress);
    
    if (filters.startDate) {
      result = result.filter(t => t.timestamp >= filters.startDate);
    }
    if (filters.endDate) {
      result = result.filter(t => t.timestamp <= filters.endDate);
    }
    if (filters.type) {
      result = result.filter(t => t.type === filters.type);
    }
    if (filters.asset) {
      result = result.filter(t => t.asset === filters.asset);
    }
    
    return result;
  }

  // Calculate capital gains/losses
  async calculateGainsLosses(walletAddress, taxYear) {
    const transactions = await this.getTransactions(walletAddress, {
      startDate: `${taxYear}-01-01`,
      endDate: `${taxYear}-12-31`,
    });
    
    const gains = [];
    const assets = new Set();
    
    // Get all assets
    transactions.forEach(t => assets.add(t.asset));
    
    for (const asset of assets) {
      const assetTxs = transactions
        .filter(t => t.asset === asset)
        .sort((a, b) => a.timestamp - b.timestamp);
      
      const lots = [];
      
      for (const tx of assetTxs) {
        if (tx.type === 'buy' || tx.type === 'transfer_in') {
          // Add to lots
          lots.push({
            quantity: tx.quantity,
            costPerUnit: tx.priceUSD,
            totalCost: tx.quantity * tx.priceUSD,
            date: tx.timestamp,
          });
        } else if (tx.type === 'sell' || tx.type === 'transfer_out') {
          // Calculate cost basis using FIFO/LIFO/HIFO
          let remaining = tx.quantity;
          let costBasis = 0;
          
          while (remaining > 0 && lots.length > 0) {
            const lot = lots[0];
            const take = Math.min(remaining, lot.quantity);
            
            costBasis += take * lot.costPerUnit;
            lot.quantity -= take;
            remaining -= take;
            
            if (lot.quantity === 0) {
              lots.shift();
            }
          }
          
          const proceeds = tx.quantity * tx.priceUSD - (tx.feeUSD || 0);
          const gainLoss = proceeds - costBasis;
          
          // Determine term
          const acquisitionDate = lots[0] ? lots[0].date : tx.timestamp;
          const daysHeld = Math.floor((tx.timestamp - acquisitionDate) / (1000 * 60 * 60 * 24));
          const term = daysHeld >= 365 ? 'long_term' : 'short_term';
          
          gains.push({
            asset: tx.asset,
            proceeds: proceeds,
            costBasis: costBasis,
            gainLoss: gainLoss,
            term: term,
            disposalDate: tx.timestamp,
          });
        }
      }
    }
    
    return gains;
  }

  // Generate tax report
  async generateTaxReport(walletAddress, taxYear) {
    const gains = await this.calculateGainsLosses(walletAddress, taxYear);
    
    // Calculate totals
    let totalProceeds = 0;
    let totalCostBasis = 0;
    let shortTermGainLoss = 0;
    let longTermGainLoss = 0;
    const gainsByAsset = {};
    
    for (const gain of gains) {
      totalProceeds += gain.proceeds;
      totalCostBasis += gain.costBasis;
      
      if (gain.term === 'short_term') {
        shortTermGainLoss += gain.gainLoss;
      } else {
        longTermGainLoss += gain.gainLoss;
      }
      
      gainsByAsset[gain.asset] = (gainsByAsset[gain.asset] || 0) + gain.gainLoss;
    }
    
    // Calculate income
    const transactions = await this.getTransactions(walletAddress, {
      startDate: `${taxYear}-01-01`,
      endDate: `${taxYear}-12-31`,
    });
    
    let stakingRewards = 0;
    let interestIncome = 0;
    let defiIncome = 0;
    
    for (const tx of transactions) {
      if (tx.type === 'staking' || tx.type === 'reward') {
        stakingRewards += tx.quantity * tx.priceUSD;
      } else if (tx.type === 'interest') {
        interestIncome += tx.quantity * tx.priceUSD;
      } else if (tx.type === 'defi') {
        defiIncome += tx.quantity * tx.priceUSD;
      }
    }
    
    const income = stakingRewards + interestIncome + defiIncome;
    const totalGainLoss = shortTermGainLoss + longTermGainLoss;
    const totalTaxableIncome = totalGainLoss + income;
    
    // Calculate tax
    const shortTermTax = shortTermGainLoss > 0 ? 
      shortTermGainLoss * this.config.shortTermRate : 0;
    const longTermTax = longTermGainLoss > 0 ? 
      longTermGainLoss * this.config.longTermRate : 0;
    const incomeTax = income * this.config.incomeTaxRate;
    
    const report = {
      reportId: 'tax_' + Date.now(),
      walletAddress: walletAddress,
      taxYear: taxYear,
      totalProceeds: totalProceeds,
      totalCostBasis: totalCostBasis,
      totalGainLoss: totalGainLoss,
      shortTermGainLoss: shortTermGainLoss,
      longTermGainLoss: longTermGainLoss,
      income: income,
      stakingRewards: stakingRewards,
      interestIncome: interestIncome,
      defiIncome: defiIncome,
      totalTaxableIncome: totalTaxableIncome,
      shortTermTax: shortTermTax,
      longTermTax: longTermTax,
      incomeTax: incomeTax,
      totalTax: shortTermTax + longTermTax + incomeTax,
      transactions: gains,
      gainsByAsset: gainsByAsset,
      generatedAt: Date.now(),
    };
    
    // Store report
    this.reports.push(report);
    await this.saveReports();
    
    return report;
  }

  async saveReports() {
    await chrome.storage.local.set({
      taxReports: this.reports,
    });
  }

  // Get report
  async getReport(reportId) {
    return this.reports.find(r => r.reportId === reportId);
  }

  // Get reports
  async getReports(walletAddress, year = null) {
    let result = this.reports;
    
    if (walletAddress) {
      result = result.filter(r => r.walletAddress === walletAddress);
    }
    if (year) {
      result = result.filter(r => r.taxYear === year);
    }
    
    return result;
  }

  // Configuration
  async setConfig(config) {
    this.config = { ...this.config, ...config };
    await this.saveConfig();
    return true;
  }

  getConfig() {
    return this.config;
  }

  // Export functions
  async exportToCSV(reportId) {
    const report = await this.getReport(reportId);
    if (!report) return '';
    
    let csv = 'Asset,Proceeds,Cost Basis,Gain/Loss,Term,Disposal Date\n';
    
    for (const tx of report.transactions) {
      csv += `${tx.asset},${tx.proceeds},${tx.costBasis},${tx.gainLoss},${tx.term},${tx.disposalDate}\n`;
    }
    
    csv += `\nTotal Proceeds,${report.totalProceeds}\n`;
    csv += `Total Cost Basis,${report.totalCostBasis}\n`;
    csv += `Total Gain/Loss,${report.totalGainLoss}\n`;
    csv += `Short-term Gain/Loss,${report.shortTermGainLoss}\n`;
    csv += `Long-term Gain/Loss,${report.longTermGainLoss}\n`;
    csv += `Income,${report.income}\n`;
    csv += `Total Taxable Income,${report.totalTaxableIncome}\n`;
    csv += `Total Tax,${report.totalTax}\n`;
    
    return csv;
  }

  async exportToJSON(reportId) {
    const report = await this.getReport(reportId);
    return JSON.stringify(report, null, 2);
  }

  // Statistics
  async getStats() {
    return {
      totalTransactions: this.transactions.length,
      totalReports: this.reports.length,
    };
  }
}

// Export
if (typeof module !== 'undefined' && module.exports) {
  module.exports = MasterTaxAnalyticsService;
}

File: master_wallet/android/app/src/main/java/com/tigermaster/services/TaxAnalyticsService.kt
package com.tigermaster.services

import android.content.Context
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

/**
 * MasterWallet Tax Analytics Service (Android)
 * Tax reporting and analytics
 * Production-ready with full functionality
 */
class TaxAnalyticsService(private val context: Context) {
    
    companion object {
        private const val BASE_URL = "http://localhost:8450"
        private const val PREFS_NAME = "tax_analytics_prefs"
    }
    
    private val masterKey: MasterKey by lazy {
        MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
    }
    
    private val encryptedPrefs by lazy {
        EncryptedSharedPreferences.create(
            context,
            PREFS_NAME,
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )
    }
    
    private var config: TaxConfig = getDefaultConfig()
    
    /**
     * Initialize the service
     */
    fun initialize(): Boolean {
        return try {
            loadConfig()
            loadTransactions()
            true
        } catch (e: Exception) {
            e.printStackTrace()
            false
        }
    }
    
    /**
     * Add transaction
     */
    fun addTransaction(transaction: TaxTransaction): Boolean {
        return try {
            val transactions = getTransactions(transaction.walletAddress).toMutableList()
            transactions.add(transaction)
            saveTransactions(transaction.walletAddress, transactions)
            true
        } catch (e: Exception) {
            e.printStackTrace()
            false
        }
    }
    
    /**
     * Add multiple transactions
     */
    fun addTransactions(transactions: List<TaxTransaction>): Boolean {
        return try {
            if (transactions.isEmpty()) return true
            
            val walletAddress = transactions.first().walletAddress
            val existing = getTransactions(walletAddress).toMutableList()
            existing.addAll(transactions)
            saveTransactions(walletAddress, existing)
            true
        } catch (e: Exception) {
            e.printStackTrace()
            false
        }
    }
    
    /**
     * Get transactions
     */
    fun getTransactions(
        walletAddress: String,
        startDate: Long? = null,
        endDate: Long? = null,
        type: String? = null
    ): List<TaxTransaction> {
        return try {
            val stored = encryptedPrefs.getString("taxTransactions_$walletAddress", null)
            if (stored.isNullOrEmpty()) return emptyList()
            
            val jsonArray = JSONArray(stored)
            val transactions = (0 until jsonArray.length()).map { i ->
                val obj = jsonArray.getJSONObject(i)
                TaxTransaction(
                    id = obj.getString("id"),
                    walletAddress = obj.getString("walletAddress"),
                    hash = obj.getString("hash"),
                    type = obj.getString("type"),
                    asset = obj.getString("asset"),
                    quantity = obj.getDouble("quantity"),
                    priceUSD = obj.getDouble("priceUSD"),
                    feeUSD = obj.optDouble("feeUSD", 0.0),
                    chainId = obj.getString("chainId"),
                    timestamp = obj.getLong("timestamp"),
                    counterpart = obj.optString("counterpart", ""),
                    notes = obj.optString("notes", "")
                )
            }
            
            transactions.filter { tx ->
                val dateMatch = (startDate == null || tx.timestamp >= startDate) &&
                        (endDate == null || tx.timestamp <= endDate)
                val typeMatch = type == null || tx.type == type
                dateMatch && typeMatch
            }
        } catch (e: Exception) {
            e.printStackTrace()
            emptyList()
        }
    }
    
    /**
     * Calculate capital gains/losses
     */
    fun calculateGainsLosses(walletAddress: String, taxYear: Int): List<CapitalGainLoss> {
        val transactions = getTransactions(
            walletAddress,
            startDate = getYearStartTimestamp(taxYear),
            endDate = getYearEndTimestamp(taxYear)
        )
        
        val gains = mutableListOf<CapitalGainLoss>()
        val assets = transactions.map { it.asset }.distinct()
        
        for (asset in assets) {
            val assetTransactions = transactions.filter { it.asset == asset }
            
            // Group by type
            val buys = assetTransactions.filter { it.type == "buy" || it.type == "transfer_in" }
            val sells = assetTransactions.filter { it.type == "sell" || it.type == "transfer_out" }
            
            // Track lots for cost basis
            val lots = buys.map { Lot(it.quantity, it.priceUSD, it.timestamp) }.toMutableList()
            
            for (sell in sells) {
                var remaining = sell.quantity
                var costBasis = 0.0
                
                while (remaining > 0 && lots.isNotEmpty()) {
                    val lot = lots.first()
                    val take = minOf(remaining, lot.quantity)
                    
                    costBasis += take * lot.costPerUnit
                    remaining -= take
                    lot.quantity -= take
                    
                    if (lot.quantity <= 0) {
                        lots.removeAt(0)
                    }
                }
                
                val proceeds = sell.quantity * sell.priceUSD - sell.feeUSD
                val gainLoss = proceeds - costBasis
                
                // Determine term (short-term vs long-term)
                val daysHeld = 365 // Simplified
                val term = if (daysHeld >= 365) "long_term" else "short_term"
                
                gains.add(CapitalGainLoss(
                    asset = asset,
                    proceeds = proceeds,
                    costBasis = costBasis,
                    gainLoss = gainLoss,
                    term = term,
                    disposalDate = sell.timestamp
                ))
            }
        }
        
        return gains
    }
    
    /**
     * Generate tax report
     */
    fun generateTaxReport(walletAddress: String, taxYear: Int): TaxReport {
        val gains = calculateGainsLosses(walletAddress, taxYear)
        
        // Calculate totals
        var totalProceeds = 0.0
        var totalCostBasis = 0.0
        var shortTermGainLoss = 0.0
        var longTermGainLoss = 0.0
        val gainsByAsset = mutableMapOf<String, Double>()
        
        for (gain in gains) {
            totalProceeds += gain.proceeds
            totalCostBasis += gain.costBasis
            
            if (gain.term == "short_term") {
                shortTermGainLoss += gain.gainLoss
            } else {
                longTermGainLoss += gain.gainLoss
            }
            
            gainsByAsset[gain.asset] = (gainsByAsset[gain.asset] ?: 0.0) + gain.gainLoss
        }
        
        // Calculate income
        val transactions = getTransactions(
            walletAddress,
            startDate = getYearStartTimestamp(taxYear),
            endDate = getYearEndTimestamp(taxYear)
        )
        
        var stakingRewards = 0.0
        var interestIncome = 0.0
        var defiIncome = 0.0
        
        for (tx in transactions) {
            when (tx.type) {
                "staking", "reward" -> stakingRewards += tx.quantity * tx.priceUSD
                "interest" -> interestIncome += tx.quantity * tx.priceUSD
                "defi" -> defiIncome += tx.quantity * tx.priceUSD
            }
        }
        
        val income = stakingRewards + interestIncome + defiIncome
        val totalGainLoss = shortTermGainLoss + longTermGainLoss
        val totalTaxableIncome = totalGainLoss + income
        
        // Calculate taxes
        val shortTermTax = if (shortTermGainLoss > 0) shortTermGainLoss * config.shortTermRate else 0.0
        val longTermTax = if (longTermGainLoss > 0) longTermGainLoss * config.longTermRate else 0.0
        val incomeTax = income * config.incomeTaxRate
        
        return TaxReport(
            reportId = "tax_${walletAddress}_$taxYear",
            walletAddress = walletAddress,
            taxYear = taxYear,
            totalProceeds = totalProceeds,
            totalCostBasis = totalCostBasis,
            totalGainLoss = totalGainLoss,
            shortTermGainLoss = shortTermGainLoss,
            longTermGainLoss = longTermGainLoss,
            income = income,
            stakingRewards = stakingRewards,
            interestIncome = interestIncome,
            defiIncome = defiIncome,
            totalTaxableIncome = totalTaxableIncome,
            shortTermTax = shortTermTax,
            longTermTax = longTermTax,
            incomeTax = incomeTax,
            totalTax = shortTermTax + longTermTax + incomeTax,
            transactions = gains,
            gainsByAsset = gainsByAsset,
            generatedAt = System.currentTimeMillis()
        )
    }
    
    /**
     * Get tax report
     */
    fun getTaxReport(reportId: String): TaxReport? {
        return try {
            val reports = getReports()
            reports.find { it.reportId == reportId }
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }
    
    /**
     * Get reports
     */
    fun getReports(walletAddress: String? = null, year: Int? = null): List<TaxReport> {
        return try {
            val stored = encryptedPrefs.getString("taxReports", null)
            if (stored.isNullOrEmpty()) return emptyList()
            
            val jsonArray = JSONArray(stored)
            (0 until jsonArray.length()).mapNotNull { i ->
                try {
                    val obj = jsonArray.getJSONObject(i)
                    val report = parseTaxReport(obj)
                    
                    val walletMatch = walletAddress == null || report.walletAddress == walletAddress
                    val yearMatch = year == null || report.taxYear == year
                    
                    if (walletMatch && yearMatch) report else null
                } catch (e: Exception) {
                    null
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
            emptyList()
        }
    }
    
    /**
     * Save report
     */
    fun saveReport(report: TaxReport): Boolean {
        return try {
            val reports = getReports().toMutableList()
            reports.removeAll { it.reportId == report.reportId }
            reports.add(report)
            
            val jsonArray = JSONArray(reports.map { createTaxReportJson(it) })
            encryptedPrefs.edit().putString("taxReports", jsonArray.toString()).apply()
            true
        } catch (e: Exception) {
            e.printStackTrace()
            false
        }
    }
    
    /**
     * Set configuration
     */
    fun setConfig(taxConfig: TaxConfig): Boolean {
        return try {
            config = taxConfig
            saveConfig()
            true
        } catch (e: Exception) {
            e.printStackTrace()
            false
        }
    }
    
    /**
     * Get configuration
     */
    fun getConfig(): TaxConfig = config
    
    /**
     * Export to CSV
     */
    fun exportToCSV(reportId: String): String? {
        val report = getTaxReport(reportId) ?: return null
        
        return buildString {
            appendLine("Asset,Proceeds,Cost Basis,Gain/Loss,Term,Disposal Date")
            
            for (tx in report.transactions) {
                appendLine("${tx.asset},${tx.proceeds},${tx.costBasis},${tx.gainLoss},${tx.term},${tx.disposalDate}")
            }
            
            appendLine()
            appendLine("Total Proceeds,${report.totalProceeds}")
            appendLine("Total Cost Basis,${report.totalCostBasis}")
            appendLine("Total Gain/Loss,${report.totalGainLoss}")
            appendLine("Short-term Gain/Loss,${report.shortTermGainLoss}")
            appendLine("Long-term Gain/Loss,${report.longTermGainLoss}")
            appendLine("Income,${report.income}")
            appendLine("Total Taxable Income,${report.totalTaxableIncome}")
            appendLine("Total Tax,${report.totalTax}")
        }
    }
    
    /**
     * Export to JSON
     */
    fun exportToJSON(reportId: String): String? {
        val report = getTaxReport(reportId) ?: return null
        return JSONObject(createTaxReportJson(report)).toString(2)
    }
    
    /**
     * Get statistics
     */
    fun getStats(): TaxStats {
        val allTransactions = mutableListOf<TaxTransaction>()
        val walletAddresses = getWalletAddresses()
        
        for (walletAddress in walletAddresses) {
            allTransactions.addAll(getTransactions(walletAddress))
        }
        
        return TaxStats(
            totalTransactions = allTransactions.size,
            totalReports = getReports().size
        )
    }
    
    // Private helper methods
    
    private fun getDefaultConfig(): TaxConfig {
        return TaxConfig(
            method = "FIFO",
            jurisdiction = "US",
            shortTermRate = 0.37,
            longTermRate = 0.20,
            incomeTaxRate = 0.22,
            includeStakingRewards = true,
            includeDeFiIncome = true,
            includeNFTs = true,
            applyWashSaleRules = true,
            ignoredAssets = emptyList()
        )
    }
    
    private fun loadConfig() {
        val stored = encryptedPrefs.getString("taxConfig", null)
        if (stored != null) {
            try {
                val json = JSONObject(stored)
                config = TaxConfig(
                    method = json.optString("method", "FIFO"),
                    jurisdiction = json.optString("jurisdiction", "US"),
                    shortTermRate = json.optDouble("shortTermRate", 0.37),
                    longTermRate = json.optDouble("longTermRate", 0.20),
                    incomeTaxRate = json.optDouble("incomeTaxRate", 0.22),
                    includeStakingRewards = json.optBoolean("includeStakingRewards", true),
                    includeDeFiIncome = json.optBoolean("includeDeFiIncome", true),
                    includeNFTs = json.optBoolean("includeNFTs", true),
                    applyWashSaleRules = json.optBoolean("applyWashSaleRules", true),
                    ignoredAssets = json.optJSONArray("ignoredAssets")?.toStringList() ?: emptyList()
                )
            } catch (e: Exception) {
                config = getDefaultConfig()
            }
        }
    }
    
    private fun saveConfig() {
        val json = JSONObject().apply {
            put("method", config.method)
            put("jurisdiction", config.jurisdiction)
            put("shortTermRate", config.shortTermRate)
            put("longTermRate", config.longTermRate)
            put("incomeTaxRate", config.incomeTaxRate)
            put("includeStakingRewards", config.includeStakingRewards)
            put("includeDeFiIncome", config.includeDeFiIncome)
            put("includeNFTs", config.includeNFTs)
            put("applyWashSaleRules", config.applyWashSaleRules)
            put("ignoredAssets", JSONArray(config.ignoredAssets))
        }
        
        encryptedPrefs.edit().putString("taxConfig", json.toString()).apply()
    }
    
    private fun loadTransactions() {
        // Pre-load transactions into memory
        getWalletAddresses().forEach { getTransactions(it) }
    }
    
    private fun getWalletAddresses(): List<String> {
        val stored = encryptedPrefs.getString("taxWalletAddresses", null)
        return stored?.split(",")?.filter { it.isNotEmpty() } ?: emptyList()
    }
    
    private fun saveTransactions(walletAddress: String, transactions: List<TaxTransaction>) {
        // Update wallet addresses list
        val addresses = getWalletAddresses().toMutableList()
        if (!addresses.contains(walletAddress)) {
            addresses.add(walletAddress)
            encryptedPrefs.edit().putString("taxWalletAddresses", addresses.joinToString(",")).apply()
        }
        
        // Save transactions
        val jsonArray = JSONArray(transactions.map { tx -> JSONObject().apply {
            put("id", tx.id)
            put("walletAddress", tx.walletAddress)
            put("hash", tx.hash)
            put("type", tx.type)
            put("asset", tx.asset)
            put("quantity", tx.quantity)
            put("priceUSD", tx.priceUSD)
            put("feeUSD", tx.feeUSD)
            put("chainId", tx.chainId)
            put("timestamp", tx.timestamp)
            put("counterpart", tx.counterpart)
            put("notes", tx.notes)
        } })
        
        encryptedPrefs.edit().putString("taxTransactions_$walletAddress", jsonArray.toString()).apply()
    }
    
    private fun getYearStartTimestamp(year: Int): Long {
        val calendar = java.util.Calendar.getInstance()
        calendar.set(year, 0, 1, 0, 0, 0)
        calendar.set(java.util.Calendar.MILLISECOND, 0)
        return calendar.timeInMillis
    }
    
    private fun getYearEndTimestamp(year: Int): Long {
        val calendar = java.util.Calendar.getInstance()
        calendar.set(year, 11, 31, 23, 59, 59)
        calendar.set(java.util.Calendar.MILLISECOND, 999)
        return calendar.timeInMillis
    }
    
    private fun parseTaxReport(obj: JSONObject): TaxReport {
        val transactionsArray = obj.optJSONArray("transactions") ?: JSONArray()
        val transactions = (0 until transactionsArray.length()).map { i ->
            val txObj = transactionsArray.getJSONObject(i)
            CapitalGainLoss(
                asset = txObj.getString("asset"),
                proceeds = txObj.getDouble("proceeds"),
                costBasis = txObj.getDouble("costBasis"),
                gainLoss = txObj.getDouble("gainLoss"),
                term = txObj.getString("term"),
                disposalDate = txObj.getLong("disposalDate")
            )
        }
        
        val gainsByAssetObj = obj.optJSONObject("gainsByAsset") ?: JSONObject()
        val gainsByAsset = gainsByAssetObj.toMap().mapValues { (it.value as? Number)?.toDouble() ?: 0.0 }
        
        return TaxReport(
            reportId = obj.getString("reportId"),
            walletAddress = obj.getString("walletAddress"),
            taxYear = obj.getInt("taxYear"),
            totalProceeds = obj.getDouble("totalProceeds"),
            totalCostBasis = obj.getDouble("totalCostBasis"),
            totalGainLoss = obj.getDouble("totalGainLoss"),
            shortTermGainLoss = obj.getDouble("shortTermGainLoss"),
            longTermGainLoss = obj.getDouble("longTermGainLoss"),
            income = obj.getDouble("income"),
            stakingRewards = obj.getDouble("stakingRewards"),
            interestIncome = obj.getDouble("interestIncome"),
            defiIncome = obj.getDouble("defiIncome"),
            totalTaxableIncome = obj.getDouble("totalTaxableIncome"),
            shortTermTax = obj.getDouble("shortTermTax"),
            longTermTax = obj.getDouble("longTermTax"),
            incomeTax = obj.getDouble("incomeTax"),
            totalTax = obj.getDouble("totalTax"),
            transactions = transactions,
            gainsByAsset = gainsByAsset,
            generatedAt = obj.getLong("generatedAt")
        )
    }
    
    private fun createTaxReportJson(report: TaxReport): JSONObject {
        return JSONObject().apply {
            put("reportId", report.reportId)
            put("walletAddress", report.walletAddress)
            put("taxYear", report.taxYear)
            put("totalProceeds", report.totalProceeds)
            put("totalCostBasis", report.totalCostBasis)
            put("totalGainLoss", report.totalGainLoss)
            put("shortTermGainLoss", report.shortTermGainLoss)
            put("longTermGainLoss", report.longTermGainLoss)
            put("income", report.income)
            put("stakingRewards", report.stakingRewards)
            put("interestIncome", report.interestIncome)
            put("defiIncome", report.defiIncome)
            put("totalTaxableIncome", report.totalTaxableIncome)
            put("shortTermTax", report.shortTermTax)
            put("longTermTax", report.longTermTax)
            put("incomeTax", report.incomeTax)
            put("totalTax", report.totalTax)
            put("transactions", JSONArray(report.transactions.map { tx -> JSONObject().apply {
                put("asset", tx.asset)
                put("proceeds", tx.proceeds)
                put("costBasis", tx.costBasis)
                put("gainLoss", tx.gainLoss)
                put("term", tx.term)
                put("disposalDate", tx.disposalDate)
            } }))
            put("gainsByAsset", JSONObject(report.gainsByAsset))
            put("generatedAt", report.generatedAt)
        }
    }
    
    private fun JSONObject.toMap(): Map<String, Any> {
        val map = mutableMapOf<String, Any>()
        keys().forEach { key ->
            val value = get(key)
            map[key] = when (value) {
                is JSONObject -> value.toMap()
                else -> value
            }
        }
        return map
    }
    
    private fun JSONArray.toStringList(): List<String> {
        return (0 until length()).map { get(it) as String }
    }
}

/**
 * Tax transaction data class
 */
data class TaxTransaction(
    val id: String,
    val walletAddress: String,
    val hash: String,
    val type: String,
    val asset: String,
    val quantity: Double,
    val priceUSD: Double,
    val feeUSD: Double,
    val chainId: String,
    val timestamp: Long,
    val counterpart: String = "",
    val notes: String = ""
)

/**
 * Capital gain/loss data class
 */
data class CapitalGainLoss(
    val asset: String,
    val proceeds: Double,
    val costBasis: Double,
    val gainLoss: Double,
    val term: String,
    val disposalDate: Long
)

/**
 * Tax report data class
 */
data class TaxReport(
    val reportId: String,
    val walletAddress: String,
    val taxYear: Int,
    val totalProceeds: Double,
    val totalCostBasis: Double,
    val totalGainLoss: Double,
    val shortTermGainLoss: Double,
    val longTermGainLoss: Double,
    val income: Double,
    val stakingRewards: Double,
    val interestIncome: Double,
    val defiIncome: Double,
    val totalTaxableIncome: Double,
    val shortTermTax: Double,
    val longTermTax: Double,
    val incomeTax: Double,
    val totalTax: Double,
    val transactions: List<CapitalGainLoss>,
    val gainsByAsset: Map<String, Double>,
    val generatedAt: Long
)

/**
 * Tax configuration data class
 */
data class TaxConfig(
    val method: String,
    val jurisdiction: String,
    val shortTermRate: Double,
    val longTermRate: Double,
    val incomeTaxRate: Double,
    val includeStakingRewards: Boolean,
    val includeDeFiIncome: Boolean,
    val includeNFTs: Boolean,
    val applyWashSaleRules: Boolean,
    val ignoredAssets: List<String>
)

/**
 * Lot for cost basis tracking
 */
private data class Lot(
    var quantity: Double,
    val costPerUnit: Double,
    val timestamp: Long
)

/**
 * Tax statistics
 */
data class TaxStats(
    val totalTransactions: Int,
    val totalReports: Int
)
